openwisp_ipam/api/views.py

```diff
@@ -28,9 +30,28 @@
     IpRequestSerializer,
     SubnetSerializer,
 )
+from .utils import FilterByOrganizationManaged
 
 IpAddress = swapper.load_model('openwisp_ipam', 'IpAddress')
 Subnet = swapper.load_model('openwisp_ipam', 'Subnet')
+Organization = swapper.load_model('openwisp_users', 'Organization')
+
+
+class DispatchOrgMixin(object):
+    def dispatch(self, *args, **kwargs):
+        self.organization = get_object_or_404(
+            self.subnet_model, pk=self.kwargs['subnet_id']
+        ).organization
+        return super().dispatch(*args, **kwargs)
+
+    def validate_membership(self, user):
+        if not (
+            user.is_superuser
+            or IsOrganizationManager.validate_membership(self, user, self.organization)
+        ):
+            raise PermissionDenied(
+                "User does not have access to the specified organization"
+            )
```

Comment:
there are several issues here: it's called DispatchOrg but you're fetching the subnet.

Can you explain what this is for?

Reply:
I'm using this to inherit in many views below, similar to how it's done in [openwisp-radius](https://github.com/openwisp/openwisp-radius/blob/a37ff581ad393dda5549d2f800edbd15b2ec3465/openwisp_radius/api/views.py#L420-L437). In an ideal case, I should be able to validate based on organization field using `organization_field='subnet__organization'`, but it doesn't work, as discussed [earlier](https://github.com/openwisp/openwisp-ipam/pull/76#pullrequestreview-532623227).

Should I remove the mixin? In that case, I'd still have to keep the function. I need help with this one.

Reply:
I think with that solution we don't even have to check if the user has access to the organization of the parent object, because the queries are combined:

1. the query which filters objects by the orgs the user is authorized to view
2. the query which looks for the specific parent object

The resulting query looks for the parent object assigned to any of the organizations the user has access to, so if the user does't have access to the org of the parent, the `exists()` method will return `False`.

The only subsequent steps we may do to generalize this is to allow looking for the organization deeper in the relationship tree (eg: if the organization field is not on the parent object but on another related model), we can keep this in mind but do it in the future when we need it.

Reply:
After looking at the code from firmware-upgrader, I created a similar mixin
```python
class SubnetMixin(OrgAPIMixin):
    def get_parent_queryset(self):
        return Subnet.objects.filter(pk=self.kwargs['subnet_id'])

    def assert_parent_exists(self):
        try:
            assert self.get_parent_queryset().exists()
        except (AssertionError, ValidationError):
            raise NotFound()

    def get_queryset(self):
        return super().get_queryset().filter(subnet=self.kwargs['subnet_id'])

    def initial(self, *args, **kwargs):
        self.assert_parent_exists()
        super().initial(*args, **kwargs)
```
But this simply doesn't prevent users from other organizations from accessing the object, not sure how to implement this properly.

Reply:
we had a discussion via IM and I gave you some hints, let me know if you have further questions.